### PR TITLE
8273392: Improve usability of stack-less exceptions due to -XX:+OmitStackTraceInFastThrow

### PR DESCRIPTION
**Pull request by simonis**
If running with `-XX:+OmitStackTraceInFastThrow` (which is the default) C2 will optimize certain "hot" implicit exceptions (i.e. AIOOBE, NullPointerExceptions,..) and replace them by a static, pre-allocated exception without any stacktrace.

However, we can actually do better. Instead of using a single, pre-allocated exception object for all methods we can let the compiler allocate specific exceptions for each compilation unit (i.e. nmethod) and fill them with at least one stack frame with the method /line-number information of the currently compiled method. If the method in question is being inlined (which often happens), we can add stackframes for all callers up to the inlining depth of the method in question.

For the attached JTreg test, we get the following exception in interpreter mode:
```
java.lang.NullPointerException: Cannot read the array length because "<parameter2>" is null
        at compiler.exceptions.StackFrameInFastThrow.throwImplicitException(StackFrameInFastThrow.java:76)
        at compiler.exceptions.StackFrameInFastThrow.level2(StackFrameInFastThrow.java:95)
        at compiler.exceptions.StackFrameInFastThrow.level1(StackFrameInFastThrow.java:99)
        at compiler.exceptions.StackFrameInFastThrow.main(StackFrameInFastThrow.java:233)
```
Once the method gets compiled with `-XX:+OmitStackTraceInFastThrow` the same exception will look as follows:
```
java.lang.NullPointerException
```
After this change, if `StackFrameInFastThrow.throwImplicitException()` will be compiled stand alone, we will get:
```
java.lang.NullPointerException
        at compiler.exceptions.StackFrameInFastThrow.throwImplicitException(StackFrameInFastThrow.java:76)
```
and if `StackFrameInFastThrow.throwImplicitException()` will be inlined into `level2()` and `level2()` into `level1()` we will get the following exception (altough we're still running with `-XX:+OmitStackTraceInFastThrow`):
```
java.lang.NullPointerException
        at compiler.exceptions.StackFrameInFastThrow.throwImplicitException(StackFrameInFastThrow.java:76)
        at compiler.exceptions.StackFrameInFastThrow.level2(StackFrameInFastThrow.java:95)
        at compiler.exceptions.StackFrameInFastThrow.level1(StackFrameInFastThrow.java:99)
```
The new functionality is guarded by `-XX:+/-StackFrameInFastThrow`, but switched on by default (I'll create a CSR for the new option once reviewers are comfortable with the change). Notice that the optimization comes at no run-time costs because all the extra work will be done at compile time.

## Implementation details

- Already the current implementation of `-XX:+OmitStackTraceInFastThrow` potentially lazy-allocates the empty singleton exceptions like AIOOBE in `ciEnv::ArrayStoreException_instance()`. With this change, if running with `-XX:+StackFrameInFastThrow` we will always allocate new exception objects and populate them with the stack frames which are statically available at compile time (see `java_lang_Throwable::fill_in_stack_trace_of_implicit_exception()`).
- Because nmethods don't act as strong GC roots, we have to create a global JNI handle for every newly generated exception to prevent GC from collecting them.
- In order to avoid a memory leak we have to release these global JNI handles once a nmethod gets unloaded. In order to achieve this, I've added a new section "implicit exceptions" to the nmethod which holds these JNI handles.
- While adding the new  "implicit exceptions" section to the corresponding stats (`print_nmethod_stats()` and printing routines (`nmethod::print()`) I realized that a previous change ([JDK-8254231: Implementation of Foreign Linker API (Incubator)](https://bugs.openjdk.java.net/browse/JDK-8254231)) had already introduced a new nmethod section ("native invokers") but missed to add it to the corresponding stats and printing routines so I've added that section as well.
- The `#ifdef COMPILER2` guards are only required to not break the `zero`/`minimal` builds.
- The JTreg test is using `-XX:PerMethodTrapLimit=0` to handle all implicit exceptions as "hot". This makes the test simpler and at the same time provokes the allocation of more implicit exceptions.
- Manually verified that the created Exception objects are freed by GC once the corresponding nmethods have been flushed.
- Manual "stress" test with a very small heap and continuous recompilation of methods with explicit exceptions to provoke GCs during compilation didn't reveal any issues.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8273392](https://bugs.openjdk.java.net/browse/JDK-8273392): Improve usability of stack-less exceptions due to -XX:+OmitStackTraceInFastThrow

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5392/head:pull/5392` \
`$ git checkout pull/5392`

Update a local copy of the PR: \
`$ git checkout pull/5392` \
`$ git pull https://git.openjdk.java.net/jdk pull/5392/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5392`

View PR using the GUI difftool: \
`$ git pr show -t 5392`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5392.diff">https://git.openjdk.java.net/jdk/pull/5392.diff</a>

</details>

### CODE REVIEW

**Comment by reinrich** [comment]
Hi Volker, Rémis,

I haven't yet looked into the details of the change but @TheRealMDoerr kindly
explained it to me. As I understood, you are using global JNI references to hold
the preallocated exceptions with partial backtrace. Backtraces seem to hold
references to the java mirrors of the holders of the methods in the backtrace
[1]. This will keep their classloaders alive and prevent classunloading. Also
the owning nmethod cannot be unloaded for the same reason. It shouldn't be to
difficult to write a test that leaks classes because of this.

> 
> _Mailing list message from [Remi Forax](mailto:forax@univ-mlv.fr) on [hotspot-dev](mailto:hotspot-dev@mail.openjdk.java.net):_
> 
> (not a reviewer so this message will not be really helpful ...)
> 
> Hi Volker,
> for me it's not an enhancement, but a bug fix, in production an exception with no stacktrace is useless and result in hours lost trying to figure out the issue


I'd agree. Even in development exceptions should have a stacktrace and
therefore, IMHO, OmitStackTraceInFastThrow should be off by default.


In my eyes exceptions are means to handle unforseen application states in an
best effort approach. Often they will be caused by bugs and the attached
stacktrace is valuable information to find them. Your enhancement limits the
stacktrace to potentially just the top frame which in many cases will not be
enough and also confusing to developers. Also I don't think that we should
optimize applications that have run into a bug.

In the related [JDK-8273563](https://bugs.openjdk.java.net/browse/JDK-8273563)
you gave the example of [Tomcat's `HttpParser::isAlpha()`
method](https://github.com/apache/tomcat/blob/26ba86cdbd40ca718e43b82e62b3eb49d004c3d6/java/org/apache/tomcat/util/http/parser/HttpParser.java#L266-L274)
method:

```java
    public static boolean isAlpha(int c) {
        try {
            return IS_ALPHA[c];
        } catch (ArrayIndexOutOfBoundsException ex) {
            return false;
        }
    }
```

There the backtrace is completely useless and can be omitted but IMHO
(stated above) this is a misuse of exceptions in the first place and should be
fixed.

Such idioms may occur in libraries that are out of maintenance (which should not
be used for security reasons) or the maintainer is not willing to accept the
fix. Therefore we could limit OmitStackTraceInFastThrow to these idioms only
where the thrown exception is caught in a local or inlined handler that ignores
it.

Maybe this is not even too difficult. If C2 compiles `isAlpha` with
OmitStackTraceInFastThrow enabled then practically everything related to
exception handling gets eliminated. It might be possible to recognize that the
IR-Node representing the preallocated exception has no uses and only if it
actually does have uses it could be replaced with an uncommon trap (which is
likely the harder part).

Cheers, Richard.

[1] Exception backtrace references java mirrors:
    https://github.com/openjdk/jdk/blob/7c9868c0b3c9bd3d305e71f91596190813cdccce/src/hotspot/share/classfile/javaClasses.cpp#L2178-L2182

File: src/hotspot/share/ci/ciEnv.cpp [hunk 1]
```diff
@@ -392,23 +393,65 @@ ciInstance* ciEnv::get_or_create_exception(jobject& handle, Symbol* name) {
...
+  InstanceKlass* ik = SystemDictionary::find_instance_klass(name, Handle(), Handle());
+  jobject objh = NULL;
+  if (ik != NULL) {
+    oop obj = ik->allocate_instance(THREAD);
+    if (!HAS_PENDING_EXCEPTION) {
+      Handle handle = Handle(THREAD, obj);
+      java_lang_Throwable::fill_in_stack_trace_of_implicit_exception(handle, gk);
+      // nmethods are no strong roots so we have to create a global JNI handle
+      // for the created exception in order to keep it alive accross GCs.
+      objh = JNIHandles::make_global(handle);
```
Comment (reinrich) [inline src/hotspot/share/ci/ciEnv.cpp:410]: The backtrace references the java mirrors corresponding to the methods in the backtrace[1]. Thereby the global JNI handle will keep their classloaders alive and prevent classunloading and also unloading of the nmethod being compiled.

[1] https://github.com/openjdk/jdk/blob/7c9868c0b3c9bd3d305e71f91596190813cdccce/src/hotspot/share/classfile/javaClasses.cpp#L2178-L2182

**Comment by reinrich** [comment]
Hi Volker,

> 
> _Mailing list message from [Volker Simonis](mailto:volker.simonis@gmail.com) on [hotspot-dev](mailto:hotspot-dev@mail.openjdk.java.net):_
> 
> Hi Richard,
> 
> thanks a lot for looking into this change.
> 
> Nmethod unloading does still work with this change, just take a look
> at the associated JTreg test which compiles and then unloads a method
> with a generated implicit exception.

Yes it works in your test because you explicitly make the compiled method not
entrant. Think of another test where a nmethod would be unloaded because the
corresponding classloader isn't reachable anymore. The change prevents this
because the loader will be kept alive by the preallocated exception if one exists.

A test with a class leak would repeatedly create a loader, c2 compile a method
with preallocated exception that was loaded by the loader and then drop the
reference to the classloader. All the loaders would be kept alive by the
preallocated exceptions.

> Once the nmethod has been
> unloaded, the global JNI handle will be released and the class can be
> unloaded as well. But I agree that it might be too late and class
> unloading shouldn't depend on unloading of all nmethods which
> reference that class. I'll have a look if I can fix that somehow.
> 
> Best regards,
> Volker
> 
> On Mon, Sep 20, 2021 at 10:53 AM Richard Reingruber
> <rrich at openjdk.java.net> wrote:

Cheers, Richard.

**Comment by simonis** [comment]
Hi Richard,

thanks one more time for pointing out the issue with the class mirrors in the backtrace which keep classes alive and potentially prevents them from being unloaded.

Fortunately, I think the solution is pretty simple. I don't think we need the backtrace at all. In the end it is just an optimization to save some space and not construct the full StackTraceElement[] right at the creation time of an exception. But the implicit exceptions which we are creating here are "nmethod-singletons" and as such I don't think we loose much if we create the  array of StackTraceElements right away instead of creating a backtrace (see my last push). The StackTraceElements only contain Strings and therefore don't keep any classes unnecessarily alive. What do you think?

And once you're on it, would you mind reviewing the whole PR :)

Thank you and best regards,
Volker

**Review by navyxliu — COMMENTED** [review]
>  for me it's not an enhancement, but a bug fix, in production an exception with no stacktrace is useless and result in hours lost trying to figure out the issue

If we treat it as a bug,  shall we remove `StackFrameInFastThrow'? we can just make this the default behavior of `OmitStackTraceInFastThrow`. 

Why `OmitStackTraceInFastThrow` isn't a c2-exclusive option?  I think it only affects c2.  That flag overrides the existing flag `StackTraceInThrowable`. Let not introduce another flag.  No one would like an exception without a pointer.

File: src/hotspot/share/classfile/javaClasses.cpp [hunk 2]
```diff
@@ -2565,6 +2568,49 @@ void java_lang_Throwable::allocate_backtrace(Handle throwable, TRAPS) {
...
+void java_lang_Throwable::allocate_fill_stack_trace_of_implicit_exception(Handle throwable, GraphKit* gk) {
+  // Fill in the stack frame(s) for an implicit exception, can cause GC
+  assert(throwable->is_a(vmClasses::Throwable_klass()), "sanity check");
+
+  JavaThread* THREAD = JavaThread::current(); // For exception macros.
+
+  InstanceKlass* ik = vmClasses::StackTraceElement_klass();
+  assert(ik != NULL, "must be loaded in 1.4+");
+
+  // Determin the number of available frames
```
Comment (navyxliu) [inline src/hotspot/share/classfile/javaClasses.cpp:2581]: typo

File: src/hotspot/share/ci/ciEnv.cpp [hunk 3]
```diff
@@ -392,23 +393,65 @@ ciInstance* ciEnv::get_or_create_exception(jobject& handle, Symbol* name) {
...
+#ifdef COMPILER2
+ciInstance* ciEnv::create_implicit_exception(Symbol* name, GraphKit* gk) {
+  VM_ENTRY_MARK;
+  InstanceKlass* ik = SystemDictionary::find_instance_klass(name, Handle(), Handle());
+  jobject objh = NULL;
+  if (ik != NULL) {
+    oop obj = ik->allocate_instance(THREAD);
+    if (!HAS_PENDING_EXCEPTION) {
+      Handle handle = Handle(THREAD, obj);
+      java_lang_Throwable::allocate_fill_stack_trace_of_implicit_exception(handle, gk);
```
Comment (navyxliu) [inline src/hotspot/share/ci/ciEnv.cpp:407]: IIHO, hotspot fills stacktrace when StackTraceInThrowable is true.

**Comment by reinrich** [comment]
Hi Volker,

> Fortunately, I think the solution is pretty simple. I don't think we need the backtrace at all. In the end it is just an optimization to save some space and not construct the full StackTraceElement[] right at the creation time of an exception. But the implicit exceptions which we are creating here are "nmethod-singletons" and as such I don't think we loose much if we create the array of StackTraceElements right away instead of creating a backtrace (see my last push). The StackTraceElements only contain Strings and therefore don't keep any classes unnecessarily alive. What do you think?

I agree. Thanks for fixing!

> And once you're on it, would you mind reviewing the whole PR :)

:)
I'll be out of office next week. Maybe I'll get to review the related https://github.com/openjdk/jdk/pull/5488 after that if needed.

Cheers, Richard.

**Comment by rose00** [comment]
To me this looks like a very clever mess.  The mess comes from the trickiness (it's a tricky problem!) and even more from forcing various parts of the system that are usually isolated to come into contact.  Adding reasons to GC during a JIT task is a smell.  Adding objects which are pieced together at compile time is a smell.  (And you can't run Java code from the JIT; it's an architectural limitation.)  Having JavaClasses talk directly to C2 GraphKit (without even a CI class between) is a smell.  Adding a new section to nmethods just to make a poorly-understood life cycle for an odd (non-Java-created) group of exceptions is a smell.

If we need a new section on nmethods it should be something more like "Java structures the JIT has made", with clearly separated concerns from the rest of the system, rather than "my very special section for an intrusive RFE".

This is not even close to being ready to integrate.

**Comment by simonis** [comment]
Hi Jon,

thanks for looking at this PR. Let me reply to your comments inline:

> To me this looks like a very clever mess. The mess comes from the trickiness (it's a tricky problem!) and even more from forcing various parts of the system that are usually isolated to come into contact. Adding reasons to GC during a JIT task is a smell. Adding objects which are pieced together at compile time is a smell. (And you can't run Java code from the JIT; it's an architectural limitation.)

I agree. But we already have all this mess today with the current implementation of `-XX:+OmitStackTraceInFastThrow` which already lazily creates empty exceptions and introduces all the problems you describe (see `ciEnv::get_or_create_exception()`). It's to a much lesser extent compared to this change, but fundamentally it is not different.

> Having JavaClasses talk directly to C2 GraphKit (without even a CI class between) is a smell. Adding a new section to nmethods just to make a poorly-understood life cycle for an odd (non-Java-created) group of exceptions is a smell.
> 
> If we need a new section on nmethods it should be something more like "Java structures the JIT has made", with clearly separated concerns from the rest of the system, rather than "my very special section for an intrusive RFE".
> 

That's a good proposal and I'm happy to work on such a solution. What do you mean exactly by ".._clearly separated concerns from the rest of the system_.."? 

> This is not even close to being ready to integrate.

As I said, I'm happy to invest more work and improve this PR based on your suggestion if there's a chance for this feature to be accepted (even if only in a heavily revised form).

But in general I think the **biggest mess** is really that users still get empty exceptions without any information at all and I think it is time to fix that. Unfortunately I can't see into the history of this code before jdk 6, but from [JDK-4292742: NullPointerException with no stack trace](https://bugs.openjdk.java.net/browse/JDK-4292742) it looks like you already worked on this issue almost 20 years ago :)

So what about removing JavaClasses' dependency on GraphKit and making the new nmethod section more generally usable as you suggested? Are there any other pain points before reconsidering this PR? Any other suggestions you like me to integrate?

Thank you and best regards,
Volker

**Comment by dean-long** [comment]
If this was the interpreter, it seems like these exception objects could go into the class constant pool as condy objects.  But because the ideal backtrace includes inlining information, it is a use case for something I'll call nmethod constant pools.  GC would probably need to scan it as strong roots (I'm not a GC expert), but it would get rid of the need for the JNI global refs.  And the exception objects could be allocated on a slow path when first needed rather than eagerly at compile time.  I have other uses in mind for nmethod constant pools, so I'm showing my bias here, but I also think a condy approach extends the system in a more elegant way.

**Pull request closed without merging**